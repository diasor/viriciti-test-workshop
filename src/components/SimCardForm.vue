<template>
  <div class="simcard-form-container">
    <h1 class="title">Testing => SimCard Form</h1>
    <form>
      <div class="simcard-form-container__element">
        <div class="simcard-form-container__label">
          <label for="simId">Sim Id:</label>
        </div>
        <input name="simId" type="text" v-model="simId" />
      </div>
      <div class="simcard-form-container__element">
        <div class="simcard-form-container__label">
          <label for="providers">Provider:</label>
        </div>
        <select
          class="custom-select"
          style="width:200px;"
          name="providers"
          id="providers"
          v-model="provider"
        >
          <option value></option>
          <option value="vodafone">Vodafone</option>
          <option value="tmobile">T-Mobile</option>
          <option value="mercedes">Another</option>
        </select>
      </div>
      <div class="simcard-form-container__element">
        <div class="simcard-form-container__label">
          <label for="status">Status:</label>
        </div>
        <select
          class="custom-select"
          style="width:200px;"
          name="status"
          id="status"
          v-model="status"
        >
          <option value></option>
          <option value="alive">alive</option>
          <option value="pending">pending</option>
          <option value="cancel">cancel</option>
        </select>
      </div>
      <div class="simcard-form-container__element">
        <div class="simcard-form-container__button">
          <button name="submitSimCard" @click.prevent="addNewSimCard()">Add SimCard</button>
        </div>
        <div class="simcard-form-container__button">
          <button
            name="addAndPrintSimCard"
            @click.prevent="addAndPrintSimCard()"
          >Add & Print SimCard</button>
        </div>
      </div>
      <div
        v-show="showNotification"
        name="successSubmit"
        class="success-submit fade-out"
      >The simcard was successfully added!</div>
      <div
        v-show="showSimCardInformation"
        name="simCardInformation"
        class="show-information fade-out"
      >SIM Card added!{{information}}</div>
    </form>
  </div>
</template>

<script>
import { mapActions } from 'vuex';

export default {
  data() {
    return {
      simCards: [],
      newSimCard: {},
      simId: '',
      provider: '',
      status: '',
      showNotification: false,
      showSimCardInformation: false,
      information: ''
    };
  },
  methods: {
    ...mapActions(['addSimCard']),

    addSimCardToState() {
      this.addSimCard({
        id: this.simId,
        provider: this.provider,
        status: this.status
      });
    },

    cleanUp() {
      this.simId = '';
      this.provider = '';
      this.status = '';
      const timeOutValue = 5000;
      if (this.showNotification) {
        setTimeout(() => (this.showNotification = false), timeOutValue);
      }
      if (this.showSimCardInformation) {
        setTimeout(() => (this.showSimCardInformation = false), timeOutValue);
      }
    },

    addNewSimCard() {
      if (this.simId === '') return;
      this.addSimCardToState();
      this.showNotification = true;
      this.cleanUp();
    },

    addAndPrintSimCard() {
      if (this.simId === '') return;

      this.information = `SIM Card Information: Id: ${this.simId}, 
        Provider: ${this.provider}, 
        Status: ${this.status}`;
      this.showSimCardInformation = true;
      this.addSimCardToState();
      this.cleanUp();
    }
  }
};
</script>
<style scoped>
.simcard-form-container {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  display: flex;
  flex-direction: column;
}
h3 {
  margin: 40px 0 0 0;
}

a {
  color: #42b983;
}
label {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  font-size: 14px;
  color: black;
  padding-top: 5px;
}

input {
  margin: 0 5px;
  width: 190px;
  height: 20px !important;
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  font-size: 14px;
}

input::placeholder {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  font-size: 14px;
  padding-top: 5px;
}

/* The container must be positioned relative: */
.custom-select {
  position: relative;
  /* font-family: Arial; */
  height: 25px;
  margin: 0 5px;
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  font-size: 14px;
}

.custom-select select {
  display: none; /*hide original SELECT element: */
}
.select-selected {
  background-color: DodgerBlue;
}

/* style the items (options), including the selected item: */
.select-items div,
.select-selected {
  color: #ffffff;
  padding: 8px 16px;
  border: 1px solid transparent;
  border-color: transparent transparent rgba(0, 0, 0, 0.1) transparent;
  cursor: pointer;
}

.select-items div:hover,
.same-as-selected {
  background-color: rgba(0, 0, 0, 0.1);
}

.simcard-form-container__element {
  display: flex;
  padding: 0.2rem 0.5rem;
  width: 50rem;
}

.simcard-form-container__label {
  display: flex;
  width: 5rem;
}

.simcard-form-container__button {
  display: flex;
  padding: 0 0.5rem 0 0;
}

.success-submit {
  display: flex;
  margin: 10px;
  padding: 10px 5px;
  height: 20px;
  width: 60vw;
  border: 1px solid green;
  border-radius: 4px;
  color: green;
  font-weight: bold;
  font-size: 14px;
}

.show-information {
  display: flex;
  margin: 10px;
  padding: 10px 5px;
  height: 30px;
  width: 60vw;
  border: 1px solid blueviolet;
  border-radius: 4px;
  color: blueviolet;
  font-weight: bold;
  font-size: 14px;
}

.fade-out {
  animation: fadeOut ease 8s;
  -webkit-animation: fadeOut ease 8s;
  -moz-animation: fadeOut ease 8s;
  -o-animation: fadeOut ease 8s;
  -ms-animation: fadeOut ease 8s;
}
@keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

@-moz-keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

@-webkit-keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

@-o-keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

@-ms-keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
</style>