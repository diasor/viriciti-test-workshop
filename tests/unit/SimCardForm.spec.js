import SimCardForm from "@/components/SimCardForm"
import { mount } from "@vue/test-utils"
import { nextTick } from "vue"

describe("testing inputs on the SimCard form", () => {
  const wrapper = mount(SimCardForm)
  expect(wrapper).toMatchSnapshot()
  const form = wrapper.find("form")

  it("finds the simId input", () =>
    expect(form.find('input[name="simId"]').exists()).toBe(true))

  it("finds the provider Select input", () =>
    expect(form.find('select[name="providers"]').exists()).toBe(true))

  it("finds the status Select input", () =>
    expect(form.find('select[name="status"]').exists()).toBe(true))

  it("finds the submit button", () =>
    expect(form.find('button[name="submitSimCard"]').exists()).toBe(true))
})

describe("setting values and submiting the SimCard form", () => {
  const wrapper = mount(SimCardForm)
  // mock the addNewSimCard
  wrapper.vm.addNewSimCard = jest.fn()
  expect(wrapper).toMatchSnapshot()
  const form = wrapper.find("form")
  const providerSelector = 'select[name="providers"]'
  const statusSelector = 'select[name="status"]'

  it("follows the user through the form and submits", async () => {
    // setting SimId input
    const inputSimId = form.find('input[name="simId"]')
    inputSimId.setValue("123456")
    expect(inputSimId.element.value).toBe("123456")

    // test that the provider select input exists, has the expected amount of options and that it has set the "vodafone" option
    const providerSelectInput = form.find(providerSelector)
    expect(providerSelectInput.findAll("option").length).toBe(4)
    form
      .find(providerSelector)
      .findAll("option")
      .at(1)
      .setSelected()
    expect(providerSelectInput.element.value).toBe("vodafone")

    // test that it was properly set
    const statusSelectInput = form.find(statusSelector)
    expect(statusSelectInput.findAll("option").length).toBe(4)
    form
      .find(statusSelector)
      .findAll("option")
      .at(1)
      .setSelected()
    expect(statusSelectInput.element.value).toBe("alive")

    // find and test the button  form.trigger("submit")
    const submitButton = form.find('button[name="submitSimCard"]')
    submitButton.trigger("click")

    // wait and test that afterwards everything is correct again
    await nextTick()
    expect(form.find('div[name="successSubmit"]').exists()).toBe(true)
  })
})
