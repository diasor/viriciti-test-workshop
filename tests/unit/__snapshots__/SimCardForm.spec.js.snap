// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div class="simcard-form-container">
  <h1 class="title">Testing Form</h1>
  <form>
    <div class="simcard-form-container__element">
      <h2>SIM Card Form</h2>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="simId">Sim Id:</label></div> <input name="simId" type="text">
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="providers">Provider:</label></div> <select name="providers" id="providers" class="custom-select" style="width: 200px;">
        <option value=""></option>
        <option value="vodafone">Vodafone</option>
        <option value="tmobile">T-Mobile</option>
        <option value="mercedes">Another</option>
      </select>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="status">Status:</label></div> <select name="status" id="status" class="custom-select" style="width: 200px;">
        <option value=""></option>
        <option value="alive">alive</option>
        <option value="pending">pending</option>
        <option value="cancel">cancel</option>
      </select>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__button"><button name="submitSimCard">Add SimCard</button></div>
      <div class="simcard-form-container__button"><button name="addAndPrintSimCard">Add &amp; Print SimCard</button></div>
    </div>
    <div name="successSubmit" class="success-submit fade-out" style="display: none;">The simcard was successfully added!</div>
    <div name="simCardInformation" class="show-information fade-out" style="display: none;">SIM Card added!</div>
  </form>
</div>
`;

exports[` 2`] = `
<div class="simcard-form-container">
  <h1 class="title">Testing Form</h1>
  <form>
    <div class="simcard-form-container__element">
      <h2>SIM Card Form</h2>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="simId">Sim Id:</label></div> <input name="simId" type="text">
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="providers">Provider:</label></div> <select name="providers" id="providers" class="custom-select" style="width: 200px;">
        <option value=""></option>
        <option value="vodafone">Vodafone</option>
        <option value="tmobile">T-Mobile</option>
        <option value="mercedes">Another</option>
      </select>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__label"><label for="status">Status:</label></div> <select name="status" id="status" class="custom-select" style="width: 200px;">
        <option value=""></option>
        <option value="alive">alive</option>
        <option value="pending">pending</option>
        <option value="cancel">cancel</option>
      </select>
    </div>
    <div class="simcard-form-container__element">
      <div class="simcard-form-container__button"><button name="submitSimCard">Add SimCard</button></div>
      <div class="simcard-form-container__button"><button name="addAndPrintSimCard">Add &amp; Print SimCard</button></div>
    </div>
    <div name="successSubmit" class="success-submit fade-out" style="display: none;">The simcard was successfully added!</div>
    <div name="simCardInformation" class="show-information fade-out" style="display: none;">SIM Card added!</div>
  </form>
</div>
`;
